integration/insert_compat_test.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/AlekSi/pointer"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+
+	"github.com/FerretDB/FerretDB/integration/setup"
+)
+
+func TestInsertCompat(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]insertCompatTestCase{
+		"InsertEmptyDocument": {
+			insert:     bson.D{},
+			resultType: nonEmptyResult,
+		},
+		"InsertIDArray": {
+			insert:     bson.D{{"_id", bson.A{"foo", "bar"}}},
+			resultType: emptyResult,
+		},
+		"InsertIDRegex": {
+			insert:     bson.D{{"_id", "^regex$"}},
+			resultType: emptyResult,
+		},
+	}
+
+	testInsertCompat(t, testCases)
+}
+
+type insertCompatTestCase struct {
+	insert        bson.D                   // required
+	resultType    compatTestCaseResultType // defaults to nonEmptyResult
+	skip          string                   // skips test if non-empty
+	skipForTigris string                   // skips test for Tigris if non-empty
+}
+
+// testInsertCompat tests insert compatibility test cases.
+func testInsertCompat(t *testing.T, testCases map[string]insertCompatTestCase) {
+	t.Helper()
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			if tc.skip != "" {
+				t.Skip(tc.skip)
+			}
+			if tc.skipForTigris != "" {
+				setup.SkipForTigrisWithReason(t, tc.skipForTigris)
+			}
+
+			t.Parallel()
+
+			ctx, targetCollections, compatCollections := setup.SetupCompat(t)
+
+			insert := tc.insert
+			require.NotNil(t, insert)
+
+			var nonEmptyResults bool
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					allDocs := FindAll(t, ctx, targetCollection)
```

Comment:
I think this is unnecessary (nor is the loop with `allDocs`). We can insert the doc once, and it's enough.